allenai/pdf-component-library
# 28683 Sample paper from JSON data

## Description

This covers the bulk of https://github.com/allenai/scholar/issues/28339

Changes in this PR:
- Created new `data` dir where our sample data lives
  - Added a JSON file with a shortened version of an annotated paper (just 3 citations)
  - Added a JSON file with citation paper data for one of the paper IDs from the sample paper
-  `Reader` updates: 
  - On page load, it kicks off async call to grab sample paper JSON data
  - Once paper data is retrieved, it "loads" annotations by grabbing the raw entities from our data and massaging it into a nicer format. We map each page number to its annotations so that the `Reader` can just loop through each page and easily get the appropriate annotations from the map. Currently citations are the only entity supported.
- `CitationPopover` updates:
  - On citation click, it kicks off an async call to grab sample citation JSON data
  - Displays 'Loading...' while we wait for data to return
  - Populates popover once we have data
  - Truncated paper abstracts to 300 chars because Ant popover placement gets messed up when the popover contents are too long
- `CitationsDemo` updates:
  - Removed hardcoded sample data
  - ~~Generates `CitationPopovers` based on new `paperAnnotated` prop~~
  - Generates `CitationPopovers` based on new `annotations` prop
- Updated `HighlightOverlayDemo`, `TextHighlightDemo`, and `ScrollToDemo` since we're using a different sample paper now
- Added new `BoundingBox` function `boundingBoxRawToScaled` to transform raw bounding box coordinates (expressed as a percentage of page size) into coordinates that make sense for our current page size 
- ~~Added new `BoundingBox` function `scaleRawBoundingBoxWithContext`- this is used to transform raw bounding box coordinates (expressed as a percentage of page size) into coordinates that make sense for our current page size, scaling, and rotation~~
- Data model updates:
  - Updated types related to entities, citations, and papers. "Raw" types represent data that is coming from our data source and non-raw types are formatted for the UI.
  - Turned `BoundingBox` `page` attribute from optional to required
  - ~~Upgraded a lot of types up to classes because we need the constructors~~ 

## Testing Plan

- Manual testing
  - New PDF loads
  - Existing functionality works as expected (zoom, rotate, outline, highlight overlay demo, text highlight demo, scroll to demo)
  - Citation popovers lazy load data, display 'Loading...' while waiting and update automatically when data comes in
  - Added console logs in various rendering/data retrieval functions to ensure that components are not rendering or grabbing new data unnecessarily 
- Ran all unit tests
